### PR TITLE
Course evaluation backend

### PR DESCRIPTION
**Pull request by mxsph**
- Adds the backend implementation for the course evaluation container
- Improves the performance of the course result view
- clean up in the `fbs.sql` file

---
- :warning: A database migration is required for testing. For this, the following tables must be added:

https://github.com/thm-mni-ii/feedbacksystem/blob/e33c8dfb2cd23e5d461ca7fdb53f87e2c666505c/ws/src/main/resources/fbs.sql#L205-L250

### CODE REVIEW

File: ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  private def calculateBonusPoints(bonusFormula: String, passedTasks: Int, toPass: Int) = {
+    val bonusPointsVariables = Map("x" -> passedTasks.toString, "y" -> toPass.toString)
+
+    try {
+      if (bonusFormula != null) formulaService.evaluate(bonusFormula, bonusPointsVariables).setScale(0, RoundingMode.HALF_UP).toInt else 0
+    } catch {
+      // If the bonus formula is invalid, the entire calculation should not fail
+      case _: Exception => 0
+    }
```
Comment (mxsph) [inline ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala:69]: - Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?  
- If the bonus formula calculation results in a negative number, you have negative bonus points, should we change this to a `0`?

Comment (Zitrone44) [reply ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala:69]: > * Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?

if easily possible i think we should change the datatype to float to allow half ponus points.

> * If the bonus formula calculation results in a negative number, you have negative bonus points, should we change this to a `0`?

I think that negative bonus points do not make sense, so we should set this to zero if negative.

Comment (mxsph) [reply ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala:69]: > if easily possible i think we should change the datatype to float to allow half ponus points.

The implementation in the backend should not be very complicated. @kimiannalena  how complex would this change be in the frontend?

> I think that negative bonus points do not make sense, so we should set this to zero if negative.

:+1:

Comment (snicki13) [reply ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala:69]: > > Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?

> if easily possible i think we should change the datatype to float to allow half ponus points.

I think that half points are not relevant for lecturers. At least I have never encountered any.
I would suggest to let the lecturer choose the method of rounding (i.e. mathematically, up or down.
As an example, in database systems we require students to solve 4 tasks to acquire one bonus point. If a student only solves 3, they do not get a point.

Comment (mxsph) [reply ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala]: > I think that half points are not relevant for lecturers. At least I have never encountered any.
> I suggest to let the lecturer choose the method of rounding (i.e. mathematically, up or down.
> As an example, in database systems we require students to solve 4 tasks to acquire one bonus point. If a student only solves 3, they do not get a point.

This is possible using functions in the formula (e.g. `FLOOR` to round the value to the nearest integer. A list of all functions can be found [here](https://github.com/uklimaschewski/EvalEx#supported-functions)). So if you want to use a rounding mode other than the default (matematic) you can, if you don't the number will automatically be rounded matematically to get an integer.

> I think that negative bonus points do not make sense, so we should set this to zero if negative.

@Zitrone44 Negative numbers are now converted to `0`.

**Review by Zitrone44 — APPROVED** [review]
:+1:

**Pull request merged**